graphium/graphium_cpp/features.cpp

```diff
@@ -0,0 +1,1474 @@
+        dtype,
+        mask_nan_style,
+        mask_nan_value,
+        num_nans);
+    if (num_nans != 0) {
+        nan_tensor_index = tensors.size()+2;
+        return;
+    }
+    tensors.push_back(std::move(edge_weights_tensor));
+    tensors.push_back(std::move(atom_features_tensor));
+    tensors.push_back(std::move(bond_features_tensor));
+    if (create_conformer_feature) {
+        at::Tensor conformer_features_tensor = get_conformer_features<T>(
+            *graph.mol,
+            already_has_Hs,
+            dtype,
+            mask_nan_style,
+            mask_nan_value,
+            num_nans,
+            smiles_string);
+        if (num_nans != 0) {
+            nan_tensor_index = tensors.size();
+            return;
+        }
+        tensors.push_back(std::move(conformer_features_tensor));
+    }
+    create_positional_features<T>(
+        graph,
+        positional_property_list,
+        dtype,
+        mask_nan_style,
+        mask_nan_value,
+        num_nans,
+        nan_tensor_index,
+        tensors);
+}
+
+std::tuple<std::vector<at::Tensor>, int64_t, int64_t> featurize_smiles(
+    const std::string& smiles_string,
+    const at::Tensor& atom_property_list_onehot,
+    const at::Tensor& atom_property_list_float,
+    bool create_conformer_feature,
+    const at::Tensor& bond_property_list,
+    const at::Tensor& positional_property_list,
+    bool duplicate_edges,
+    bool add_self_loop,
+    bool explicit_H,
+    bool use_bonds_weights,
+    bool offset_carbon,
+    int dtype_int,
+    int mask_nan_style_int,
+    double mask_nan_value) {
+
+    GraphData graph = read_graph(smiles_string, explicit_H);
+
+    const size_t edge_coo_count = 2*graph.num_bonds + (add_self_loop ? graph.num_atoms : 0);
+    std::unique_ptr<int64_t[]> edge_index(new int64_t[2*edge_coo_count]);
+    for (size_t i = 0; i < graph.num_bonds; ++i) {
+        // PyG has all directed edge begin indices followed by all end indices.
+        edge_index[2*i] = graph.bonds[i].beginAtomIdx;
+        edge_index[2*i+1] = graph.bonds[i].endAtomIdx;
+        edge_index[2*i + edge_coo_count] = graph.bonds[i].endAtomIdx;
+        edge_index[2*i+1 + edge_coo_count] = graph.bonds[i].beginAtomIdx;
+    }
+    if (add_self_loop) {
+        for (size_t i = 0; i < graph.num_atoms; ++i) {
+            edge_index[2*graph.num_bonds + i] = i;
+            edge_index[2*graph.num_bonds + i + edge_coo_count] = i;
+        }
+    }
+    int64_t edge_coo_dims[2] = { int64_t(2), int64_t(edge_coo_count) };
+    at::Tensor edge_coo_tensor = torch_tensor_from_array<int64_t>(std::move(edge_index), edge_coo_dims, 2, c10::ScalarType::Long);
+
+    std::vector<at::Tensor> tensors;
+    tensors.push_back(std::move(edge_coo_tensor));
+    c10::ScalarType dtype = c10::ScalarType(dtype_int);
+    MaskNaNStyle mask_nan_style = MaskNaNStyle(mask_nan_style_int);
+    int64_t num_nans = 0;
+    int64_t nan_tensor_index = -1;
+    if (dtype == c10::ScalarType::Half) {
```

Comment:
Is it necessary to repeat the exact same if/else if statement 3 times, except for the `<int16_t>` or `<float>` or `<double>`? Can you use a variable for that?

My C++ is a bit rusty, it has been years.

Reply:
C++ templates (e.g. the `<int16_t>` or `<float>` or `<double>`) need to be able to be evaluated at compile-time, whereas the if statements are evaluated at runtime.  These are the if statements that check the type at runtime once, so that it can be encoded in the template, and then it doesn't need to be evaluated at runtime again.